service-base/src/main/java/org/eclipse/hono/service/command/Command.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.service.command;
+
+import java.net.HttpURLConnection;
+import java.util.Map;
+
+import io.vertx.core.buffer.Buffer;
+import org.apache.qpid.proton.amqp.Binary;
+import org.apache.qpid.proton.amqp.messaging.ApplicationProperties;
+import org.apache.qpid.proton.amqp.messaging.Data;
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.util.MessageHelper;
+
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonDelivery;
+import io.vertx.proton.ProtonHelper;
+
+/**
+ * Represents a Command and Control API command on the receiver side (adapter).
+ * <p>
+ * Received command includes the message data and all application properties.
+ */
+public class Command {
+
+    private Object correlationId;
+    private String replyAddress;
+    private String contentType;
+    private Buffer messageData;
+    private Map<String, Object> applicationProperties;
+    private CommandResponder responder;
+
+    Command(final CommandResponder responder, final Message message) {
```

Comment:
IMHO it would be better to keep the `message` as a member instead of copying some values into separate members.
It would make the class much smaller, and keeps it more flexible. Or would you think of serializing this class and thus rather prefer the `basic` types inside this class?

Reply:
Yes, I understand that.
My point is: it looks like a lot (not all) parts of a message is "copied out" to the members, which might be just fine - it just feels strange that e.g. the application properties are fully copied, some other values as well, but others not. And by this defining a class that is something like "half a message".
But thinking about it now it might be what we need, if we somehow persist a command in the future (which I think could come quite soon). Then it will be fine to copy out only the releveant parts.

So how about explaining this a little in the class javadoc, so others won't ask the same questions? If it is written that this intentionally pulls part of the message to ease persistence, I would not have asked myself this question.